c7n/tags.py

```diff
@@ -913,6 +914,78 @@ def process_resource_set(self, client, resource_set, tag_keys):
             client.untag_resources, ResourceARNList=arns, TagKeys=tag_keys)
 
 
+class UniversalTagRename(Action):
+    """Rename an existing tag key to a new value.
+
+    :example:
+
+    rename Application, and Bap to App, if a resource has both of the old keys
+    then we'll use the value specified by Application, which is based on the
+    order of values of old_keys.
+
+        .. code-block :: yaml
+
+            policies:
+            - name: rename-tags-example
+              resource: aws.log-group
+              filters:
+                - or:
+                  - "tag:Bap": present
+                  - "tag:Application": present
+              actions:
+                - type: rename-tag
+                  old_keys: [Application, Bap]
+                  new_key: App
+    """
+    schema = utils.type_schema(
+        'rename-tag',
+        old_keys={'type': 'array', 'items': {'type': 'string'}},
+        old_key={'type': 'string'},
+        new_key={'type': 'string'},
+    )
+
+    permissions = UniversalTag.permissions + UniversalUntag.permissions
+
+    def validate(self):
+        if 'old_keys' not in self.data and 'old_key' not in self.data:
+            raise PolicyValidationError(
+                f"{self.manager.ctx.policy.name}:{self.type} 'old_keys' or 'old_key' required")
+
+    def process(self, resources):
+        old_keys = set(self.data.get('old_keys', ()))
+        if 'old_key' in self.data:
+            old_keys.add(self.data['old_key'])
+
+        # Collect by distinct tag value, and old_key value to minimize api calls
+        # via bulk api usage on add and remove.
+        old_key_resources = {}
+        values_resources = {}
+
+        # sort tags using ordering of old_keys
+        def key_func(a):
+            if a['Key'] in old_keys:
+                return self.data['old_keys'].index(a['Key'])
+            return 50
+
+        for r in resources:
+            found = False
+            for t in sorted(r.get('Tags', ()), key=key_func):
```

Comment:
Good call this covers the ordering and avoids duplicate tag calls :+1: 

I think we could get a policy-ordered list of relevant old tags with something like this too:

```python
all_tags = {t['Key']: t['Value'] for t in r.get('Tags', [])}
old_tags = {k: all_tags[k] for k in old_keys if k in all_tags}
```

Which seems like it could avoid a level of nesting and the need for a custom key func, but I don't _think_ it's a clear enough win to justify changing.